### PR TITLE
particle filter update

### PR DESCRIPTION
**Pull request by aleslamitz**
Updating particle filter, adding rejuvenation and extra

### CODE REVIEW

File: tensorflow_probability/python/experimental/mcmc/sequential_monte_carlo_kernel.py [hunk 1]
```diff
@@ -115,19 +118,35 @@ def _dummy_indices_like(indices):
...
+                  particles,
+                  indices,
+                  log_weights,
+                  extra,
+                  seed
+                  ):
+  return extra
+
+
+def identity(state, new_particles, new_indices, log_weights, extra, step):
```
Comment (jburnim) [inline tensorflow_probability/python/experimental/mcmc/sequential_monte_carlo_kernel.py:148]: It looks like this function is unused.  Please remove.

File: tensorflow_probability/python/experimental/mcmc/sequential_monte_carlo_kernel.py [hunk 2]
```diff
@@ -115,19 +118,35 @@ def _dummy_indices_like(indices):
       indices_shape)
 
 
-def log_ess_from_log_weights(log_weights):
+def log_ess_from_log_weights(log_weights, particles_dim=0):
   """Computes log-ESS estimate from log-weights along axis=0."""
```
Comment (jburnim) [inline tensorflow_probability/python/experimental/mcmc/sequential_monte_carlo_kernel.py:122]: "along axis=particles_dim"

File: tensorflow_probability/python/experimental/mcmc/sequential_monte_carlo_kernel.py [hunk 3]
```diff
@@ -49,6 +49,9 @@ class WeightedParticles(collections.namedtuple(
       `exp(reduce_logsumexp(log_weights, axis=0)) == 1.`. These must be used in
       conjunction with `particles` to compute expectations under the target
       distribution.
+    extra: a (structure of) Tensor(s) each of shape
+      `concat([[num_particles, b1, ..., bN], event_shape])`, where `event_shape`
```
Comment (jburnim) [inline tensorflow_probability/python/experimental/mcmc/sequential_monte_carlo_kernel.py:53]: I think `num_particles` shouldn't be in the shape here.  (And could the comment mention that this represents global state of the sampling process that is *not* associated with individual particles?)

File: tensorflow_probability/python/experimental/mcmc/weighted_resampling_test.py [hunk 4]
```diff
@@ -299,6 +299,46 @@ def resample_with_target_distribution(self):
...
+        stacked_particles, stacked_log_weights,
+        resample_fn=resample_systematic,
+        particles_dim=1,
+        seed=test_util.test_seed(sampler_type='stateless'))
+
+    self.assertAllMeansClose(new_particles,
+                             [20., 20.],
+                             axis=1,
+                             atol=1e-2)
+    self.assertAllClose(
```
Comment (jburnim) [inline tensorflow_probability/python/experimental/mcmc/weighted_resampling_test.py:319]: This might be a more clear as:
```
self.assertAllClose(
        tf.reduce_sum(tf.nn.softmax(new_log_weights) * new_particles, axis=1),
        [20., 20.],
        atol=1e-2)
```

File: tensorflow_probability/python/experimental/mcmc/weighted_resampling_test.py [hunk 5]
```diff
@@ -299,6 +299,46 @@ def resample_with_target_distribution(self):
...
+        resample_fn=resample_systematic,
+        particles_dim=1,
+        target_log_weights=poisson.Poisson(30).log_prob(particles),
+        seed=test_util.test_seed(sampler_type='stateless'))
+    self.assertAllMeansClose(new_particles,
+                             [20., 20.],
+                             axis=1,
+                             atol=1e-2)
+
+    self.assertAllClose(
```
Comment (jburnim) [inline tensorflow_probability/python/experimental/mcmc/weighted_resampling_test.py:337]: This might be more clear as:
```
    self.assertAllClose(
        tf.reduce_sum(tf.nn.softmax(new_log_weights) * new_particles, axis=1),
        [30., 30.],
        atol=1.)
```

File: tensorflow_probability/python/experimental/mcmc/weighted_resampling_test.py [hunk 6]
```diff
@@ -299,6 +299,46 @@ def resample_with_target_distribution(self):
         tf.reduce_sum(tf.nn.softmax(new_log_weights) * new_particles),
         30., atol=1.)
 
+  def test_with_target_distribution_dim_one(self):
+    particles = np.linspace(0., 500., num=2500, dtype=np.float32)
+    stacked_particles = np.stack([particles, particles], axis=0)
```
Comment (jburnim) [inline tensorflow_probability/python/experimental/mcmc/weighted_resampling_test.py:304]: It may be worth making the two different batch dimensions contain different particles -- e.g., something like:
```
stacked_particles = np.stack([
    np.linspace(0., 500., num=2500, dtype=np.float32),
    np.linspace(0.17, 433., num=2500, dtype=np.float32),
], axis=0)
```

File: tensorflow_probability/python/experimental/mcmc/weighted_resampling_test.py [hunk 7]
```diff
@@ -299,6 +299,46 @@ def resample_with_target_distribution(self):
...
+        40.,
+        atol=1e-2)
+
+    # Reweight the resampled particles to target a Poisson(30.) distribution.
+    new_particles, _, new_log_weights = resample(
+        stacked_particles,
+        stacked_log_weights,
+        resample_fn=resample_systematic,
+        particles_dim=1,
+        target_log_weights=poisson.Poisson(30).log_prob(particles),
```
Comment (jburnim) [inline tensorflow_probability/python/experimental/mcmc/weighted_resampling_test.py:330]: particles -> stacked_particles

File: tensorflow_probability/python/experimental/mcmc/particle_filter.py [hunk 8]
```diff
@@ -359,88 +490,122 @@ def particle_filter(observations,
...
+    initial_state_prior = batch_broadcast.BatchBroadcast(initial_state_prior, batch_shape[:dim] + [num_samples] + batch_shape[dim:])
+    return initial_state_prior.sample(seed=seed)
 
 
 def _particle_filter_initial_weighted_particles(observations,
                                                 observation_fn,
                                                 initial_state_prior,
                                                 initial_state_proposal,
-                                                num_particles,
+                                                num_inner_particles,
```
Comment (jburnim) [inline tensorflow_probability/python/experimental/mcmc/particle_filter.py:552]: I think this argument should still be called `num_particles`.

Callers of this function might have different particle dimensions for different nested SMCs / particle filters -- e.g., "inner" and "outer".  But from the perspective of this function, I don't think the concept of "inner particles" makes sense -- we're just creating the initial state for some particle filter that has `num_particles` particles, indexed by dimension `particles_dim`.

**Review by jburnim — COMMENTED** [review]
This is looking great!

I would like to chat about how we want to handle `extra` and rejuvenation when we next meet.  Until then, could we simplify this first PR to just make the changes for?:

 1. Handling `particles_dim` in weighted_resampling.py, sequential_monte_carlo_kernel.py, and particle_filter.py.

 2. Introducing the `sequential_monte_carlo` function in `particle_filter.py`.

(In particular, we would leave out adding `extra` to `WeightedParticles`, as well as any new code for doing rejuvenation in `particle_filter.py`.) 

(We could still replace the `rejuvenation_kernel_fn=None` argument with `rejuvenation_fn=None` and `rejuvenation_criterion_fn=None`, but let's leave these arguments unused until the next PR.)

I think we could get the simplified PR in quickly, and then be able to focus on rejuvenation, exta, and SMC^2.

File: tensorflow_probability/python/experimental/mcmc/particle_filter.py [hunk 9]
```diff
@@ -273,19 +273,126 @@ def observation_fn(_, state):
...
+          lambda seed_state_results: trace_criterion_fn(  # pylint: disable=g-long-lambda
+              *seed_state_results[1:])),
+      static_trace_allocation_size=static_trace_allocation_size,
+      parallel_iterations=parallel_iterations)
+
+  if trace_criterion_fn is never_trace:
+    # Return results from just the final step.
+    traced_results = trace_fn(*final_seed_state_result[1:])
+
+    return traced_results
```
Comment (jburnim) [inline tensorflow_probability/python/experimental/mcmc/particle_filter.py:380]: This line needs to be un-indented -- so we return `trace_results` whether or not `trace_criterion_fn is never_trace`.

File: tensorflow_probability/python/experimental/mcmc/sequential_monte_carlo_kernel.py [hunk 10]
```diff
@@ -145,6 +148,7 @@ def __init__(self,
                resample_fn=weighted_resampling.resample_systematic,
                resample_criterion_fn=ess_below_threshold,
                unbiased_gradients=True,
+               particles_dim=0,
```
Comment (jburnim) [inline tensorflow_probability/python/experimental/mcmc/sequential_monte_carlo_kernel.py:151]: It looks like the `bootstrap` method also needs to be updated to handle `particles_dim`.

That is, instead of initializing `incremental_log_marginal_likelihood` and `accumulated_log_marginal_likelihood` to:
```
batch_zeros = tf.zeros(
    ps.shape(init_state.log_weights)[1:],
    dtype=init_state.log_weights.dtype)
```
we need to initialize them to something like:
```
        particles_shape = ps.shape(init_state.log_weights)
        weights_shape = ps.concat([
            particles_shape[:self.particles_dim],
            particles_shape[self.particles_dim+1:]
        ], axis=0)
        batch_zeros = tf.zeros(
            weights_shape, dtype=init_state.log_weights.dtype)
```


To verify that we have done this correctly, could you also please add a version of `test_batch_of_filters` in `particle_filter_test.py` with a `particles_dim` of 1 or 2?

File: tensorflow_probability/python/experimental/mcmc/particle_filter.py [hunk 11]
```diff
@@ -474,7 +585,8 @@ def propose_and_update_log_weights_fn(step, state, seed=None):
           particles=proposed_particles,
           log_weights=log_weights + _compute_observation_log_weights(
```
Comment (jburnim) [inline tensorflow_probability/python/experimental/mcmc/particle_filter.py:586]: It looks like `_compute_observation_log_weights` needs to be updated to add a `particles_dim` argument.

And that function will have to do something like:
```
observation = tf.expand_dims(observation, axis=particles_dim)
```
so that the computed log-weights have the right shape.

File: tensorflow_probability/python/experimental/mcmc/particle_filter.py [hunk 12]
```diff
@@ -344,7 +461,8 @@ def particle_filter(observations,
 
   init_seed, loop_seed = samplers.split_seed(seed, salt='particle_filter')
   with tf.name_scope(name or 'particle_filter'):
-    num_observation_steps = ps.size0(tf.nest.flatten(observations)[0])
```
Comment (jburnim) [inline tensorflow_probability/python/experimental/mcmc/particle_filter.py:347]: I think this should still be getting the size of the 0-th dimension.  I don't believe the shape of the observations is affected by particles_dim.

File: tensorflow_probability/python/experimental/mcmc/particle_filter.py [hunk 13]
```diff
@@ -514,30 +660,42 @@ def _compute_observation_log_weights(step,
         ps.equal(step % num_transitions_per_observation, 0))
     observation_idx = step // num_transitions_per_observation
     observation = tf.nest.map_structure(
-        lambda x, step=step: tf.gather(x, observation_idx), observations)
+        lambda x, step=step: tf.gather(x,
+                                       observation_idx,
+                                       axis=particles_dim),
```
Comment (jburnim) [inline tensorflow_probability/python/experimental/mcmc/particle_filter.py:665]: I think this line should be unchanged -- because the shape of `observations` is not affected by particles_dim.

File: tensorflow_probability/python/experimental/mcmc/particle_filter.py [hunk 14]
```diff
@@ -514,30 +660,42 @@ def _compute_observation_log_weights(step,
...
+        observations)
+
+    # Expand dimensions of observation at the particles_dim
+    observation = tf.expand_dims(observation, axis=particles_dim)
```
Comment (jburnim) [inline tensorflow_probability/python/experimental/mcmc/particle_filter.py:669]: NOTE: The goal of this expand_dims is something like:

 - `observations` has shape like `[num_steps] + batch_shape`.

 - so `observation` has shape like `batch_shape`.

 - particles has shape like `batch_shape_1 + [num_particles] + batch_shape_2 + event_shape`

 - `observation_fn(step, particles)` is a distribution with batch shape like `batch_shape_1 + [num_particles] + batch_shape_2` and event shape `event_shape`.

 - When particles_dim=0, we can just apply `observation_fn(step, particles).log_prob(observations)`, because the `log_prob` is required to broadcast the observations of shape `batch_shape + event_shape` up to `[num_particles] + batch_shape + event_shape`.

 - But when particles_dim>0, this won't work.  Instead, we have to expand_dims on `observations` so it has shape like `batch_shape_1 + [1] + batch_shape_2 + [event_shape`.  This will allow the `log_prob` to broadcast `observations` up to shape `batch_shape_1 + [num_steps] + batch_shape_2 + [event_shape]`, so that the `log_prob` can run and will return something with shape `batch_shape_1 + [num_steps] + batch_shape_2`.

**Comment by jburnim** [comment]
It looks like there are merge conflicts we need to resolve before we can submit this PR.


To resolve these merge conflicts, I recommend either:

 1. Create a brand new fork and a new branch in that fork, and copy over the six modified files (particle_filter.py, particle_filter_test.py, sequential_monte_carlo_kernel.py, sequential_monte_carlo_kernel_test.py, weighted_resampling.py, and weighted_resampling_test.py) and create a new PR.

 2. Or:

    a. First get the main branch of our fork to match the main branch of https://github.com/tensorflow/probability , by following the steps under Option B at https://scribu.net/blog/resetting-your-github-fork.html .

    b. Then get this branch in a clean state.  It may work to merge the updated main branch into this branch.  But it might also be easier to delete and recreate the branch with a single commit containing your changes to the six modified files.

**Comment by aleslamitz** [comment]
https://github.com/tensorflow/probability/pull/1771
